src/webgpu/api/validation/encoding/cmds/dynamic_render_state.spec.ts

```diff
@@ -0,0 +1,299 @@
+    pass.endPass();
+
+    this.expectValidationError(() => {
+      encoder.finish();
+    }, !success);
+  }
+
+  testScissorCall(
+    success: boolean | 'type-error',
+    s: ScissorCall,
+    attachmentSize: GPUExtent3D = { width: 1, height: 1, depth: 1 }
+  ) {
+    const attachment = this.device.createTexture({
+      format: 'rgba8unorm',
+      size: attachmentSize,
+      usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+    });
+
+    const encoder = this.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: attachment.createView(),
+          loadValue: 'load',
+        },
+      ],
+    });
+    if (success === 'type-error') {
+      this.shouldThrow('TypeError', () => {
+        pass.setScissorRect(s.x, s.y, s.w, s.h);
+      })
+    } else {
+      pass.setScissorRect(s.x, s.y, s.w, s.h);
+      pass.endPass();
+
+      this.expectValidationError(() => {
+        encoder.finish();
+      }, !success);
+    }
+  }
+
+  createDummyRenderPassEncoder() : {encoder: GPUCommandEncoder, pass : GPURenderPassEncoder}{
+    const attachment = this.device.createTexture({
+      format: 'rgba8unorm',
+      size: [1, 1, 1],
+      usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+    });
+
+    const encoder = this.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: attachment.createView(),
+          loadValue: 'load',
+        },
+      ],
+    });
+
+    return {encoder, pass};
+  }
+
+}
+
+export const g = makeTestGroup(F);
+
+g.test('setViewport,x_y_width_height_nonnegative')
+  .desc('Test that the parameters of setViewport to define the box must be non-negative.')
+  .params([
+    // Control case: everything to 0 is ok, covers the empty viewport case.
+    { x: 0, y: 0, w: 0, h: 0 },
+
+    // Test -1
+    { x: -1, y: 0, w: 0, h: 0 },
+    { x: 0, y: -1, w: 0, h: 0 },
+    { x: 0, y: 0, w: -1, h: 0 },
+    { x: 0, y: 0, w: 0, h: -1 },
+
+    // TODO Test -0 (it should be valid) but can't be tested because the harness complains about duplicate parameters.
+    // TODO Test the first value smaller than -0
+  ])
+  .fn(t => {
+    const { x, y, w, h } = t.params;
+    const success = x >= 0 && y >= 0 && w >= 0 && h >= 0;
+    t.testViewportCall(success, { x, y, w, h, minDepth: 0, maxDepth: 1 });
+  });
+
+g.test('setViewport,xy_rect_contained_in_attachment')
+  .desc(
+    'Test that the rectangle defined by x, y, width, height must be contained in the attachments'
+  )
+  .params(
+    params()
+      .combine([
+        { attachmentWidth: 3, attachmentHeight: 5 },
+        { attachmentWidth: 5, attachmentHeight: 3 },
+        { attachmentWidth: 1024, attachmentHeight: 1 },
+        { attachmentWidth: 1, attachmentHeight: 1024 },
+      ])
+      .combine([
+        // Control case: a full viewport is valid.
+        { dx: 0, dy: 0, dw: 0, dh: 0 },
```

Comment:
optional nit: consider phrasing as "left right top bottom" instead of "dx dy dw dh", it's easier to understand how it relates to the attachment size.